### PR TITLE
feature(fix): function parameter tracking

### LINKED ISSUE #5480: TypeError: Cannot read properties of null (reading 'render')
### Rollup Version

4.16.0

### Operating System (or Browser)

Ubuntu 20.04.6 LTS / MacOS Sonoma 14.4.1

### Node Version (if applicable)

18.20.2

### Link To Reproduction

https://github.com/elchininet/shadow-dom-selector/pull/55

### Expected Behaviour

Build should success as it succeeded on previous versions.

### Actual Behaviour

Build fails with an error coming from `ConditionalExpression.render`:

```bash
at ConditionalExpression.render (/xxx/node_modules/rollup/dist/shared/rollup.js:11185:24)
    at VariableDeclarator.render (/xxx/node_modules/rollup/dist/shared/rollup.js:12038:18)
    at VariableDeclaration.render (/xxx/node_modules/rollup/dist/shared/rollup.js:13724:28)
    at renderStatementList (/xxx/node_modules/rollup/dist/shared/rollup.js:9106:25)
    at BlockStatement.render (/xxx/node_modules/rollup/dist/shared/rollup.js:9272:13)
    at FunctionDeclaration.render (/xxx/node_modules/rollup/dist/shared/rollup.js:6348:23)
    at ExportNamedDeclaration.render (/xxx/node_modules/rollup/dist/shared/rollup.js:11436:30)
    at renderStatementList (/xxx/node_modules/rollup/dist/shared/rollup.js:9095:35)
    at Program.render (/xxx/node_modules/rollup/dist/shared/rollup.js:12880:13)
    at Module.render (/xxx/node_modules/rollup/dist/shared/rollup.js:15641:18)
```

I tried to reproduce it in Repl without success, but seeing that it is coming from `ConditionalExpression.render` that was recently refactored this could give a clue to the devs.

### PR DESCRIPTION
**Pull request by liuly0322**
This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [x] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

- resolves #5480 

### Description

This PR is to point out a problem with 4.16, and discuss if we should revert it first if the work to do is heavy.

Unused node should not be included, otherwise:

```
logicalExpression1 -- logicalExpression2 -- LogicalExpression3
                                         -- LogicalExpression4
                   -- logicalExpression5 -- LogicalExpression6
                                         -- LogicalExpression7
```

In the first visit, assuming no optimization info for node1, so node2 and node5 will both get included. and node6 may get not included.

However, if in the second visit, node5 is excluded, becaused it is still mark `included`, in render stage renderer will try to render node5 and find node6 or node 7 error.

This PR can fix #5480. (no tests because the condition is a bit harsh and I have not reproduced a smaller one)

While it is fixable, It actually reveals: we need a way to revert `include`. This can be done in two ways to make sure no problems:

1. explicitly set `included=false` in every `include` function that may only include some children (or `deoptimizePath` like in the PR, it actually makes no difference because `deoptimizeCache` will request one `include`, but place it in `include` may be better because there may be other functions causing children included change (in the future?) and request a `include`) 
2. in render function, we don't use `included` to check, but only uses `usedBranch` or other stuffs like that. I guess it's not how it is designed.

This PR only achives a part of the first way. I'll be off for next 12 or more hours, so if it's too heavy to do, maybe we should revert version 4.16 ...

### CODE REVIEW

**Comment by liuly0322** [comment]
> Not sure I fully get the underlying connections, but at first glance I feel very uneasy about reverting includes. It has been a reliable behavior for tree-shaking that things only ever get included, and it could even make the algorithm non-deterministic if we change that. In the worst case, we include something unnecessarily, but I would rather have that than have the algorithm not reliably converge.

Wow, yes, the real problem here is not don't include anything unnecessary, but make sure the children include has the latest state. I'll rethink it this way.

**Comment by lukastaegert** [comment]
I think the problem is that branch resolution should never go from analyzed to not-analyzed but rather from analyzed to unoptimized as it used to be. That ensures that it stays deterministic. We CAN re-analyze the resolution, but if the result is that then the other branch is picked, we should instead go for not-analyzed IMO. Not sure I will manage to do a rollback now, but maybe it is ok to leave the issue open for a day.

**Comment by liuly0322** [comment]
> I think the problem is that branch resolution should never go from analyzed to not-analyzed but rather from analyzed to unoptimized as it used to be. That ensures that it stays deterministic. We CAN re-analyze the resolution, but if the result is that then the other branch is picked, we should instead go for not-analyzed IMO. Not sure I will manage to do a rollback now, but maybe it is ok to leave the issue open for a day.

I think the problem is not restricted to picking the other branch, a short example (still I have not reproduced, so only an example):

```javascript
function foo(l, r) {
  // l and r may be a complex expression
  console.log(l && r)
}

foo(l, r)
```

If in the first pass, l and r are both unknown, both l and r will be included. However, the value of r is unknown does not mean nothing happens inside, there is probably something happens in r, and request a tree-shaking to update `include`.
If in the second pass, function parameters updated, so only l is picked. the tree-shaking update of `include` of r is then missing!

I force pushed a commit, to keep the idea with "include", make sure if one branch is included, then it's included in next tree-shakings.

Also I think it may be fine to update `branchResolutionAnalyzed`, since it only indicated "isUsedBranchCache"?

**Comment by lukastaegert** [comment]
I'm still not super happy about this. The idea that an unused branch could be included seems wrong to me. Why not go back to the original implementation for now where something that is unknown at one point remains unknown? We would still keep most of the function parameter optimization but this part would be on more solid footing?

**Comment by lukastaegert** [comment]
Looking at your example, why is the function call not known during the first pass when the function is included? After all, it is only included because of the function call.

**Comment by liuly0322** [comment]
> Looking at your example, why is the function call not known during the first pass when the function is included? After all, it is only included because of the function call.

I believe is a cache problem, the function gets included because of the function call, [expression cache](https://github.com/rollup/rollup/blob/master/test/form/samples/tree-shake-literal-parameter/expression-cache/main.js) (which fails if I don't set `isBranchResolutionAnalyzed=false`) but before that it seems it has an cached value of unoptimized.

Other one test with a conditional expression will fail too, but yes, we still have most of the optimizations, and don't include unused branch seems reasonable, so I'll remove `isBranchResolutionAnalyzed=false` and update tests now.

It would be good to know why we will have a unknown expression cache for conditional/logical expression before include and if we can clear it before first include. But maybe not in this PR?

**Comment by lukastaegert** [comment]
Maybe we just revert the change to IfStatement for now as well, i.e.

```js
deoptimizeCache(): void {
  this.testValue = UnknownValue;
}
```

as it was before, and we can then think about a better long-term approach after this is released?

**Comment by liuly0322** [comment]
> Maybe we just revert the change to IfStatement for now as well, i.e.
> 
> ```js
> deoptimizeCache(): void {
>   this.testValue = UnknownValue;
> }
> ```
> 
> as it was before, and we can then think about a better long-term approach after this is released?

It almost discards all of the parameter optimizations, so maybe we revert main branch to 4.15 code, and move all commits after that to a new development branch?

**Comment by lukastaegert** [comment]
I created #5847 to revert for now. While this is not as nice as rewinding master back to an earlier version, it ensures that master is kept in a releasable state while maintaining the release history and avoiding force pushes.

**Comment by liuly0322** [comment]
So I just reverted #5487 in this PR, we can collect and fix bugs for function parameters tracking here.

Some known issues:

1. cache logic
2. https://github.com/rollup/rollup/pull/5443#issuecomment-2072149468
3. Maybe we first make it an experimental feature to see if it's stable for some time?
   > Would you be able to add an option separate from treeshake for this feature (like experimental_treeshake?), which is disabled by default

**Comment by lukastaegert** [comment]
Ok, I have some thoughts about the problems we were facing. As you correctly saw, the function parameters are collected far too late. The problem is not the collection, though: This happens on `applyDeoptimizations` for the CallExpression. The problem is that the parameter values are updated imperatively only on `include`, which is far too late.

Here is a proposal how this can be fixed:
Instead of imperatively storing the argument values on the ParameterVariable on FunctionBase.include, we could instead directly store the argument values on FunctionBase.deoptimizeArgumentsOnInteractionAtPath on the ParameterVariables. This process would
- directly consolidate new values with the existing values for each parameter
- make sure that we can only change from no value to one value to unknown value
- deoptimize caches if we transition to "unknown", see below

Beyond that, any functions like `ParameterVariable.getLiteralValueAtPath` would get the current parameter value but also store the `origin` parameter in its cache so that we the known value changes, we can reactively deoptimize any other part that depends on the value.

As the CallExpression should be traversed before the function body is traversed, the call values should usually be present. However, if we overlook anything and the ParameterVariable is queried before it has a call value (or before we know there is a usage that we do not track), we could just set the value to Unknown and keep it like that. But I do not expect this to happen.

What do you think?

**Comment by liuly0322** [comment]
So I update my code:

1. reverted changes to `IfStatement`, `ConditionalExpression` and `LogicalExpression`, and other changes to cache design.
2. refactor `ParameterVariable` and `FunctionBase`, now `isReassigned` also indicates if it's a bottom state. (I found old implementation bottom state confusing). Also, bottom state means "no optimization", so we can use `EMPTY_ARRAY` again.

The deoptimize cache handler and `CallExpression.hasEffects` may not be perfect, so I call it a draft.

**Comment by liuly0322** [comment]
As code gets smaller, I can take more time to see where the problems might be and I found one. [REPL](http://localhost:5173/repl/?local=true&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKGEpJTIwJTdCJTVDbiUyMCUyMGFyZ3VtZW50cyU1QjAlNUQlMjAlM0QlMjAwJTVDbiUyMCUyMGV2YWwoJTVDJTIyYXJndW1lbnRzJTVCMCU1RCUyMCUzRCUyMDAlNUMlMjIpJTVDbiUyMCUyMGlmJTIwKGEpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGNvbnNvbGUubG9nKDEpJTVDbiUyMCUyMCU3RCUyMGVsc2UlMjAlN0IlNUNuJTIwJTIwJTIwJTIwY29uc29sZS5sb2coMCklNUNuJTIwJTIwJTdEJTVDbiU3RCU1Q24lNUNuZm9vKDEpJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=)

While I assumed functions parameters/arguments to be static, It seems not. Luckily it's the only thing I found break the rule. `Function.prototype.arguments` is not standard and deprecated. (Since we definately cannot access a local variable without name (or most minifiers shouldn't work), `arguments` is the only hole. And `arguments` itself is a local variable, we must access it from name (or eval))

From my perspective, we need to make `eval` and `arguments` deoptimize function parameters...?

**Comment by liuly0322** [comment]
Well, it seems `eval` is an old problem, [repl](https://rollupjs.org/repl/?version=4.16.4&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKGEpJTIwJTdCJTVDbiUyMCUyMGV2YWwoJTVDJTIyYXJndW1lbnRzJTVCMCU1RCU1QjAlNUQlMjAlM0QlMjAwJTVDJTIyKSU1Q24lN0QlNUNuJTVDbmxldCUyMGElMjAlM0QlMjAlNUIxJTVEJTVDbiU1Q25mb28oYSklNUNuJTVDbmlmJTIwKGElNUIwJTVEKSUyMCU3QiU1Q24lMjAlMjBjb25zb2xlLmxvZygxKSU1Q24lN0QlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMm91dHB1dCUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTdEJTJDJTIydHJlZXNoYWtlJTIyJTNBdHJ1ZSU3RCU3RA==), so it's ok we leave `eval` alone?

Fixing `arguments` is surprisingly easy, I'll push a commit with fix and test.

**Comment by lukastaegert** [comment]
> Why this.applyDeoptimizations() is not placed at the entry of the function, but the exit?

Honestly, I am not sure, I guess the order just did not matter. Try switching the order around, if no test fails it should be fine.

> Making an Data-flow analysis optimization converge usually has two ways:
> In every loop, iterate over all nodes. that's what we used to do.
use a work-list algorithm. Push node-b to work-list if node-a updates and node-b depends on node-a.

Interesting that you mention this. In the far future, I would envision Rollup to transfer most of the tree-shaking logic to what you call a work-list algorithm. So one would only have a single full pass and then maintain a list of "dirty" nodes that need to be re-evaluated. For myself, I always called this a "reactive" algorithm, but work-list probably fits better. If done right, this should improve performance quite a bit. However, it would require to know all dependencies of all nodes and this feels quite complicated to implement.

I think I will have a look at your changes by tomorrow.

> From my perspective, we need to make eval and arguments deoptimize function parameters...?

`eval` is usually ignored by Rollup as you get a big warning when it is used. It can just break in unexpected ways as variables are renamed dynamically. However, using `arguments` should deoptimize parameters. There IS logic to track the `arguments` variable to make sure all parameters are included when it is used, look for `argumentsVariable` in the code.

**Comment by lukastaegert** [comment]
Ah but you are right, reassignments of parameters themselves should not be tracked in `arguments`.

**Comment by liuly0322** [comment]
> Try switching the order around, if no test fails it should be fine.

No test fails, I find https://github.com/rollup/rollup/pull/4148 related. I will have a look today.

> If done right, this should improve performance quite a bit. However, it would require to know all dependencies of all nodes and this feels quite complicated to implement.

Really cool, in LLVM IR it's easy because it is trivial to build a use-def chain for `op src1 src2 dst`, but not sure how it can be done with AST.

> Ah but you are right, reassignments of parameters themselves should not be tracked in arguments.

I'm just a little worried because I'm not familiar with how people would use rollup, and maybe there are some legacy code depends on this behavior. The fix is relatively easy, though.

```typescript
export default class FunctionNode extends FunctionBase {
                super.include(context, includeChildrenRecursively);
                this.id?.include();
                const hasArguments = this.scope.argumentsVariable.included;
+               if (hasArguments) {
+                       this.deoptimizeParameters();
+               }
                // ...
}
```

**Review by lukastaegert — APPROVED** [review]
This looks quite good now. I wonder if we should try another release :)

**Comment by liuly0322** [comment]
> This looks quite good now. I wonder if we should try another release :)

Thanks, I think it's definitely worth trying.

File: src/ast/nodes/Identifier.ts [hunk 1]
```diff
@@ -60,10 +60,12 @@ export default class Identifier extends NodeBase implements PatternNode {
 		}
 	}
 
+	private isReferenceVariable = false;
 	bind(): void {
 		if (!this.variable && isReference(this, this.parent as NodeWithFieldDefinition)) {
 			this.variable = this.scope.findVariable(this.name);
 			this.variable.addReference(this);
+			this.isReferenceVariable = true;
```
[the remark this answers is not in the document]
Comment (lukastaegert) [reply src/ast/nodes/Identifier.ts:68]: Ah, I see, thanks

**Comment by kleinfreund** [comment]
Apologies for not providing a concise bug report in the form of a new issue, but I can’t really share the actual reproducible code (private repository). If I find some time, I’ll try working on a proper bug report.

I’m pretty sure 4.17.0 has the same regression I’ve seen in rollup >=4.16.0 <4.16.4. The following kind of code (I obfuscated the real code but kept, hopefully, the syntax and logic sufficiently accurately):

```js
import { exportA, exportB } from 'package'

export function fn(a, b, c) {
  return async (d, e) => {
    const { result } = await exportB(d)
    if (!b.prop1?.prop2?.prop3) {
      c.error = new Error('', { cause: { prop: a.fn('string') } })
      c.dispatchEvent(new ErrorEvent('error', { error: c.error }))

      return true
    }
    if (!result) {
      if (!exportA) {
        c.error = new Error('', { cause: { prop: a.fn('string') } })
        c.dispatchEvent(new ErrorEvent('error', { error: c.error }))
      }

      return true
    }
  }
}
```

becomes, functionally, in the output:

```js
import { exportA, exportB } from 'package'

export function fn(a, b, c) {
  return async (d, e) => {
    await exportB(d)
    c.error = new Error('', { cause: { prop: a.fn('string') } })
    c.dispatchEvent(new ErrorEvent('error', { error: c.error }))
    return true
  }
}
```

I think the second `if` statement isn’t relevant. What seems to happen is that `if (!b.prop1?.prop2?.prop3) {` gets "optimized" away making the first `return` statement return from the async function unconditionally and then, following this optimization logically, the second `if` statement doesn't matter because it's dead code.

**Comment by lukastaegert** [comment]
If you want to help us track this down, we need to know what all places where the function `fn` is called look like, especially where it would take the second branch:
- what do the parameters look like?
- what does the line of code look like where the call is made?

For some reason, it can neither track the call nor can I detect that it "lost track" of the function. It could also be of interest if the function is passed to other functions, assigned to object properties or global variables etc.

**Comment by kleinfreund** [comment]
Of course, but first, let me start by saying that 4.17.2 (i.e. https://github.com/rollup/rollup/pull/5503) fixes the issue I describe in https://github.com/rollup/rollup/pull/5483#issuecomment-2082159384.

> we need to know what all places where the function `fn` is called look like, especially where it would take the second branch:
> 
> * what do the parameters look like?
> * what does the line of code look like where the call is made?

`fn` is called exactly once and the parameters are provided as `fn(...args)`. The line containing the call is the sole body of an arrow function:

```js
      prop: (...args) => {
        return [fn(...args)]
      },
```

> It could also be of interest if the function is passed to other functions, assigned to object properties or global variables etc.

In the file that contains the sole call of `fn`, it is accessed through a direct, named import from the file `fn` is exported from. It’s not referenced or called anywhere else.

Looking at https://github.com/rollup/rollup/pull/5503/files, it seems my example might be exactly what’s fixed in that PR. If you think there could be more to this, I’d be happy to provide more information of course.

**Comment by lukastaegert** [comment]
Thanks for taking the time to respond, this gives me some confidence that @liuly0322 got to the root of the problem here. Lets see if anything else pops up in the next days, but so far it is looking quite good!

**Pull request merged**